solvers/python_pgd.py

```diff
@@ -0,0 +1,86 @@
+from benchopt import BaseSolver, safe_import_context
+
+with safe_import_context() as import_ctx:
+    import numpy as np
+    from scipy import sparse
+    from numpy.linalg import norm
+    prox_l1sorted = import_ctx.import_from('utils', 'prox_l1sorted')
+    prox_slope = import_ctx.import_from('utils', 'prox_slope')
+
+
+class Solver(BaseSolver):
+    name = 'Slope-PGD'  # proximal gradient
+    stopping_strategy = "callback"
+
+    # any parameter defined here is accessible as a class attribute
+    parameters = {'prox': ['prox_l1sorted', 'prox_slope']}
+    references = [
+        'I. Daubechies, M. Defrise and C. De Mol, '
+        '"An iterative thresholding algorithm for linear inverse problems '
+        'with a sparsity constraint", Comm. Pure Appl. Math., '
+        'vol. 57, pp. 1413-1457, no. 11, Wiley Online Library (2004)',
+        'A. Beck and M. Teboulle, "A fast iterative shrinkage-thresholding '
+        'algorithm for linear inverse problems", SIAM J. Imaging Sci., '
+        'vol. 2, no. 1, pp. 183-202 (2009)'
+    ]
+
+    def skip(self, X, y, alphas):
+        fit_intercept = False
+        # XXX - not implemented but not too complicated to implement
+        if fit_intercept:
+            return True, f"{self.name} does not handle fit_intercept"
+
+        return False, None
+
+    def set_objective(self, X, y, alphas):
+        self.X, self.y, self.lambdas = X, y, alphas
+        self.fit_intercept = False
+
+    def st(self, w, mu):
+        w -= np.clip(w, -mu, mu)
+        return w
+
+    def get_result(self):
+        return self.w
+
+    def compute_lipschitz_constant(self):
+        if not sparse.issparse(self.X):
+            L = np.linalg.norm(self.X, ord=2) ** 2
+        else:
+            L = sparse.linalg.svds(self.X, k=1)[1][0] ** 2
+        return L
+
+    def run(self, callback):
+        fit_intercept = False
+        n_samples, n_features = self.X.shape
+        R = self.y.copy()
+        w = np.zeros(n_features)
+        intercept = 0.0
+
+        z = w.copy()
+        t = 1
+
+        if sparse.issparse(self.X):
```

Comment:
here you could do:
`L = self.compute_lipschitz_constant()`
to avoid code duplication.
as is done in https://github.com/benchopt/benchmark_lasso/blob/main/solvers/python_pgd.py